integrated_channels/canvas/client.py

```diff
@@ -78,22 +94,85 @@ def course_update_endpoint(canvas_base_url, course_id):
+            self.session,
+            edx_course_id
+        )
 
-        # step 2: upload image_url and any other details
-        self._update_course_details(created_course_id, serialized_data)
+        if not located_course:
+            # Course does not exist: Create the course
+            status_code, response_text = self._post(self.course_create_url, serialized_data)
+            created_course_id = json.loads(response_text)['id']
+
+            # step 2: upload image_url and any other details
+            self._update_course_details(created_course_id, course_details)
+        else:
+            workflow_state = located_course['workflow_state']
+            if workflow_state.lower() == 'deleted':
+                LOGGER.error(
+                    generate_formatted_log(
+                        'canvas',
+                        self.enterprise_configuration.enterprise_customer.uuid,
+                        None,
+                        edx_course_id,
+                        'Course with integration_id = {edx_course_id} found in deleted state, '
+                        'not attempting to create/update'.format(
+                            edx_course_id=edx_course_id,
+                        ),
+                    )
+                )
+                status_code = 200
+                response_text = MESSAGE_WHEN_COURSE_WAS_DELETED
+            else:
+                # 'unpublished', 'completed' or 'available' cases
+                LOGGER.warning(
+                    generate_formatted_log(
+                        'canvas',
+                        self.enterprise_configuration.enterprise_customer.uuid,
+                        None,
+                        edx_course_id,
+                        'Course with canvas_id = {course_id},'
+                        'integration_id = {edx_course_id} found in workflow_state={workflow_state},'
+                        ' attempting to update instead of creating it'.format(
+                            course_id=located_course["id"],
+                            edx_course_id=edx_course_id,
+                            workflow_state=workflow_state,
+                        ),
+                    )
+                )
+                status_code, response_text = self._update_course_details(
+                    located_course['id'],
+                    course_details,
+                )
 
         return status_code, response_text
 
     def update_content_metadata(self, serialized_data):
         self._create_session()
 
         integration_id = self._extract_integration_id(serialized_data)
-        course_id = self._get_course_id_from_integration_id(integration_id)
+        course_id = CanvasUtil.get_course_id_from_edx_course_id(
+            self.enterprise_configuration,
+            self.session,
+            integration_id,
+        )
 
         url = CanvasAPIClient.course_update_endpoint(
             self.enterprise_configuration.canvas_base_url,
```

Comment:
I recommend creating a getter for the `canvas_base_url` and ensuring that it appends a trailing slash to the end  of the value.  We definitely get burned on this from time to time, especially with SAPSF.

Reply:
sure! I will take a look after pr to see if we manage that or not. but good callout, I think stripping makes sense to be safe to manual errors, e.g.

Reply:
I can even do a micro pr for that

Reply:
ever better, (in another PR), I think we want to use urllib.parse. 

```
>>> import urllib.parse
>>> urllib.parse.urljoin('http://test1','/abc')
'http://test1/abc'
>>> urllib.parse.urljoin('http://test1/','/abc')
'http://test1/abc'
>>> urllib.parse.urljoin('http://test1/','abc')
'http://test1/abc'
>>> urllib.parse.urljoin('http://test1','abc')
'http://test1/abc'
```